draft.md

```diff
@@ -0,0 +1,81 @@
+# Draft Ipld Unixfs Spec
+
+## Basic Structure
+
+   - Some sort of header that indicates that this a directory and included a version number.  The header could also have fields to give additional information on the meaning of the extended attributes.
+
+  - CBOR Map
+    - Key: CBOR Byte or Text String: File Name
+    - Value: CBOR Array of:
```

Comment:
I am really concerned having attributes being represented as an array:
- it hampers extensibility - we are certainly forgetting something today, which means almost certainly yet another optional map in the future, and then which one is which?
- It makes parsing difficult when represented as JSON

Why not just a map? With all keys being [pre-agreed upon multicodec-style](https://github.com/multiformats/multicodec/blob/master/table.csv): i.e. it must exist in one of the centralized spec tables in order to be recognized by anyone

We can still declare *some* of the keys as mandatory, and it is at the discretion of gateways/nodes/etc to decide what to do with "obviously malformed" blocks. We already have this with protobuf/unixfs: if one uploads a link-block with only "type 2" fields, and no "data" - everything rejects it.

Reply:
CBOR defines sorting logic for canonicalizing maps, but not for arrays, and a canonical representation for unixfs directory should be a must.

Reply:
> Why not just a map?

I covered the reasons below in the notes section.

@ehmry I don't see how your comment applies.

Reply:
> I covered the reasons below

You only mentioned "more compact" which is a dubious goal - we would be saving couple bytes per link. The separation of "optional" from "mandatory" into 2 maps seems more wasteful than the saving from the array...

Reply:
I am not tied to this idea.  The amount of space saving is something that can be calculated once we determine what the keys will be if we use a map.

Reply:
@kevina I assumed you meant an array of [key, value] tuples. CBOR is supposed to be schema-less and ordered values seem like schema.

Reply:
Even if we go with a map for directory entries certain keys will be required in order for the directory entry to be well defined so that is also a schema in a way.

Reply:
Well, assigning integer keys to the spec attributes would order them in the same way and just be one byte of overhead for each attribute in a map.